### ISSUE TITLE
import/first does not consider export..from

### ISSUE THREAD

**Issue by Kovensky** [benmosher/eslint-plugin-import#801]
Currently, `import/first` seems to require that any `import` comes before any `export`, even if the `export` is an `export..from`. Reordering them, in this case, could be a potential logic error.

Though, arguably, if the module is capable of exporting, it should also be made side-effect free...

**Comment by ljharb** [comment]
Why would that make it harder to tell?

You'd read the `export { members, comment }`, and then having already read the two imports, you'd know they were being re-exported.

Why is this situation any different from `const foo = 3; export { foo }; const bar = 4; export { bar }` (which should be `const foo = 3; const bar = 4; export { foo, bar }`)?

**Comment by Kovensky** [comment]
This is a small snippet of a larger file with many reexports.

Overall, I don't see any reason, anywhere, ever, for considering `export {}` as code. It should never trigger use before define or other similar lint warnings. If you want to consolidate your exports, then an export-specific rule should be used.

**Comment by ljharb** [comment]
Of course an export is code, as is an import - exports are conceptually like a return.

It seems like the sole purpose of the rule isn't something you want enforced; I'd suggest disabling the rule in that case.

> Reordering them, in this case, could be a potential logic error.

Regrouping your import/export pairs such that imports are first, and then exports, can't possibly change the meaning of the code (assuming that the imports are all in the same order), so I'm not sure what you mean here.

**Comment by Kovensky** [comment]
@ljharb `export { foo } from 'bar'`

**Comment by Andarist** [comment]
I've also bumped into this while linking packages generated by rollup for local testing purposes (yes - I could change rule settings to ignore those directories, but that's not the point). I agree with @Jessidhia that because reexports are imports they should just be allowed there. From my POV the purpose of this rule is to avoid confusion around this:
```js
console.log('second')
import './i-will-log-first'
```

Reexports on the top won't confuse people like the given snippet can. 

I've created a quick PR for this https://github.com/benmosher/eslint-plugin-import/pull/1522 , merge it or close it I guess, your call - just wanted to bring up my 2 cents into this.
